File: src/main/java/itmo/is/project/service/trade/TradePolicyService.java
package itmo.is.project.service.trade;

import itmo.is.project.dto.trade.policy.TradePolicyDto;
import itmo.is.project.dto.trade.policy.UpdateTradePolicyRequest;
import itmo.is.project.mapper.trade.TradePolicyMapper;
import itmo.is.project.model.resource.Resource;
import itmo.is.project.model.trade.TradePolicy;
import itmo.is.project.repository.trade.TradePolicyRepository;
import itmo.is.project.service.resource.ResourceService;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class TradePolicyService {
    private final TradePolicyRepository tradePolicyRepository;
    private final TradePolicyMapper tradePolicyMapper;
    private final ResourceService resourceService;

    public Page<TradePolicyDto> getAll(Pageable pageable) {
        return tradePolicyRepository.findAll(pageable).map(tradePolicyMapper::toDto);
    }

    public Page<TradePolicyDto> getAllStationBuys(Pageable pageable) {
        return tradePolicyRepository.findAllByStationBuysTrue(pageable).map(tradePolicyMapper::toDto);
    }

    public Page<TradePolicyDto> getAllStationSells(Pageable pageable) {
        return tradePolicyRepository.findAllByStationSellsTrue(pageable).map(tradePolicyMapper::toDto);
    }

    public TradePolicyDto getByResourceId(Integer resourceId) {
        return tradePolicyRepository.findById(resourceId).map(tradePolicyMapper::toDto).orElseThrow(() ->
                new EntityNotFoundException("Trade policy not found with resource id: " + resourceId)
        );
    }

    @Transactional
    public TradePolicyDto updateByResourceId(Integer resourceId, UpdateTradePolicyRequest request) {
        Resource resource = resourceService.getResourceById(resourceId);
        TradePolicy tradePolicy = tradePolicyMapper.toEntity(request);
        tradePolicy.setResource(resource);
        tradePolicy = tradePolicyRepository.save(tradePolicy);
        return tradePolicyMapper.toDto(tradePolicy);
    }
}

File: src/main/kotlin/itmo/isproject/service/trade/TradePolicyService.kt
package itmo.isproject.service.trade

import io.github.oshai.kotlinlogging.KotlinLogging
import io.github.oshai.kotlinlogging.withLoggingContext
import itmo.isproject.dto.trade.policy.TradePolicyDto
import itmo.isproject.dto.trade.policy.UpdateTradePolicyRequest
import itmo.isproject.mapper.trade.TradePolicyMapper
import itmo.isproject.repository.trade.TradePolicyRepository
import itmo.isproject.service.resource.ResourceService
import jakarta.persistence.EntityNotFoundException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

private val logger = KotlinLogging.logger {}

@Service
class TradePolicyService(
    private val tradePolicyRepository: TradePolicyRepository,
    private val tradePolicyMapper: TradePolicyMapper,
    private val resourceService: ResourceService
) {

    fun getAll(pageable: Pageable): Page<TradePolicyDto> {
        withLoggingContext("page" to pageable.pageNumber.toString()) {
            logger.debug { "Fetching all trade policies" }
        }
        return tradePolicyRepository.findAll(pageable).map { tradePolicyMapper.toDto(it) }
    }

    fun getAllStationBuys(pageable: Pageable): Page<TradePolicyDto> {
        withLoggingContext("page" to pageable.pageNumber.toString()) {
            logger.debug { "Fetching all station buys trade policies" }
        }
        return tradePolicyRepository.findAllByStationBuysTrue(pageable).map { tradePolicyMapper.toDto(it) }
    }

    fun getAllStationSells(pageable: Pageable): Page<TradePolicyDto> {
        withLoggingContext("page" to pageable.pageNumber.toString()) {
            logger.debug { "Fetching all station sells trade policies" }
        }
        return tradePolicyRepository.findAllByStationSellsTrue(pageable).map { tradePolicyMapper.toDto(it) }
    }

    fun getByResourceId(resourceId: Int): TradePolicyDto {
        withLoggingContext("resourceId" to resourceId.toString()) {
            logger.debug { "Fetching trade policy by resourceId" }
        }
        return tradePolicyRepository.findByIdOrNull(resourceId)?.let { tradePolicyMapper.toDto(it) }
            ?: throw EntityNotFoundException("Trade policy not found with resource id: $resourceId")
    }

    @Transactional
    fun updateByResourceId(resourceId: Int, request: UpdateTradePolicyRequest): TradePolicyDto {
        withLoggingContext(
            "resourceId" to resourceId.toString(),
            "stationBuys" to (request.stationBuys?.toString() ?: "null"),
            "stationSells" to (request.stationSells?.toString() ?: "null")
        ) {
            logger.info { "Updating trade policy" }
        }
        val resource = resourceService.getResourceById(resourceId)
        var tradePolicy = tradePolicyMapper.toEntity(request)
        tradePolicy.resource = resource
        tradePolicy = tradePolicyRepository.save(tradePolicy)
        withLoggingContext(
            "resourceId" to resourceId.toString(),
            "buyPrice" to (tradePolicy.purchasePrice?.toString() ?: "null"),
            "sellPrice" to (tradePolicy.sellPrice?.toString() ?: "null")
        ) {
            logger.info { "Trade policy updated" }
        }
        return tradePolicyMapper.toDto(tradePolicy)
    }
}
